enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,461 @@
+
+## Motivation
+
+MicroShift is built for edge devices, which commonly use peripherals
+like sensors and adapters.
+Currently, MicroShift does not provide any solution to pass these peripherals
+through to Pods without requiring elevated privileges.
+We want to provide a solution for attaching simple devices (not requiring any special
+initialization) to Pods without potentially compromising on security.
+
+### User Stories
+
+* As a MicroShift device administrator, I want to attach simple devices to Pods
+  without elevating privileges (e.g. mounting `/dev`, adding extraneous capabilities).
+
+### Goals
+
+* Include functionality of
+  [Generic Device Plugin](https://github.com/squat/generic-device-plugin)
+  in MicroShift
+
+### Non-Goals
+
+* Adapt Generic Device Plugin (e.g. create Operator) for OpenShift or other Kubernetes distribution
+
+## Proposal
+
+* Import Generic Device Plugin code into MicroShift binary
+* Integrate the Generic Device Plugin's configuration into MicroShift config
+  and add a toggle option
+
+### Workflow Description
+
+**User** is a human user responsible for setting up and managing Edge Devices.
+
+1. The user enables the Generic Device Plugin using MicroShift configuration
+1. Optionally, the user modifies the configuration of Generic Device Plugin, 
+   for example changes device groups and paths
+1. The user (re)starts MicroShift
+1. MicroShift starts the Generic Device Plugin controller which creates plugin
+   goroutines for each device group.
+1. Plugin goroutines detect the devices and report them with the Kubelet
+1. Kubelet updates the Node's status with new devices for scheduler to be aware of Node's capabilities
+1. User deploys a workload utilizing some device
+   (could be at any point, it won't be scheduled/started until the device is available)
+
+In order for workload to get the device attached, the **container** specification
+must include the device in its limits. Below is an example of StatefulSet which
+references a device in a container's limits.
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: WORKLOAD
+  namespace: NAMESPACE
+spec:
+  template:
+    spec:
+      containers:
+      - name: CONTAINER
+        resources:
+          limits:
+            device.microshift.io/serial: "1"
+```
+
+### API Extensions
+
+By including Generic Device Plugin (GDP) as code import, MicroShift configuration file
+will grow to accommodate the GDP's settings.
+
+Example config (based on default settings of GDP with changed `domain` and added `status`):
+```yaml
+genericDevicePlugin:
+  status: { Enabled , Disabled }
+  domain: device.microshift.io
+  devices:
+    - name: serial
+      groups:
+        - paths:
+            - path: /dev/ttyUSB*
+        - paths:
+            - path: /dev/ttyACM*
+              mountPath: /dev/ttyACM0
+    - name: video
+      groups:
+        - paths:
+            - path: /dev/video0
+    - name: fuse
+      groups:
+        - count: 10
```

Comment:
How does this count get determined ?

Reply:
It's whatever user sets. If it's set to 10, then in node's status you'll see 10 allocatable devices.

Reply:
@pmtk Does it mean it can be set to any random number and it is still acceptable ? Are there no limits or anything like that ? If user tries to set something like 1000(not sure if this would be feasible or not though) and attach so many devices would it not impact anything with respect to performance ?

Reply:
I couldn't find any limit set in the Generic Device Plugin.
There's not guarantee here I think. You could try to run 1000 host level processes at once accessing /dev/fuse, so I think everything is possible

Reply:
okay...